### ISSUE TITLE
Give administrators UI to edit rules

### ISSUE THREAD

**Issue by jasonaowen** [OpenTechStrategies/psm#274]
Each state will have different requirements around provider enrollments: things like what credentials are acceptable, what screening needs to take place, and so on. We are using a rules engine to manage all that, but our current workflow for changing those rules is not very user-friendly.

Present a UI to administrators to allow them to edit those rules at runtime.

**Comment by jasonaowen** [comment]
(Taken in part from https://github.com/OpenTechStrategies/psm/issues/262#issuecomment-315403537)

Currently, we are using jBPM 5.4 and Drools 5.5. jBPM and Drools are closely related to Guvnor, and it appears that there was some support built in to use Guvnor (see [the WebSphere deployment guide (.doc file download link)](https://github.com/OpenTechStrategies/psm/raw/33c780bc948ed0abf4aa35d2848622d4865b49f9/psm-app/docs/Websphere%20Deployment%20Guide.doc)). The related version of Guvnor is incompatible with WildFly 10; we will need to upgrade jBPM to get to a deployable version of Guvnor (now the KIE Workbench).

**Comment by jasonaowen** [comment]
@brainwane and I talked through where I'm at with this, and there was some useful background and some worthwhile conclusions.

[Stack Overflow](https://stackoverflow.com/a/9736570/25637) has a useful description of the difference between jBPM and Drools:

> jBPM is a workflow engine. Drools is a rule engine.

We've spent a lot of time looking at the rules (see #262) we execute with Drools. We haven't spent as much time looking at how we use jBPM.

There is one workflow defined: [EnrollmentProcess](https://github.com/OpenTechStrategies/psm/blob/7f8d25276b92a5831eb14bc71d38616722eded32/psm-app/cms-business-process/src/main/resources/EnrollmentProcess.bpmn), written in [Business Process Model Notation](https://en.wikipedia.org/wiki/Business_Process_Model_and_Notation). Alongside that jBPM-executable code is a diagram I believe was generated from it:

![EnrollmentProcess BPMN diagram](https://raw.githubusercontent.com/OpenTechStrategies/psm/7f8d25276b92a5831eb14bc71d38616722eded32/psm-app/cms-business-process/src/main/resources/EnrollmentProcess.png)

A single workflow is not very many workflows; is this worth bringing in such a heavyweight dependency as jBPM? Could we jettison jBPM and just encode this workflow in Java? @brainwane and I talked ourselves into believing that different states might well need different workflows. One point in favor of that is the [SIRS review](http://www.dhs.state.mn.us/main/idcplg?IdcService=GET_DYNAMIC_CONVERSION&RevisionSelectionMethod=LatestReleased&dDocName=id_008921#sirs) that moderate- and high-risk providers must undergo appears to be Minnesota-specific. So, we should keep jBPM.

(If that conclusion is wrong, if the expected differences between states' enrollment processes is small, please let us know!)

That still leaves me with a daunting upgrade. The APIs that changed from jBPM 5.4.0 to 6.0.0 are significant, at least the way we're using them, and I haven't been able to find much documentation about it.

One approach I begun today and will continue to try tomorrow is to revert some of the changes made in support of WebSphere. When the application was targeting JBoss AS, the jBPM engine was run in its own thread and the code that called it communicated over the network using [Apache MINA](https://en.wikipedia.org/wiki/Apache_MINA). As part of the migration to WebSphere, they converted to a local model: each interaction with jBPM seems to spin up just the part of jBPM it needs (I think). I don't know the reasoning behind switching away from a long-lived thread or why they built this local model; I also don't know if there are other approaches that might be better (can jBPM be deployed to WildFly standalone, as its own jar/war/ear?). Hopefully, in the process of converting back to thread+network, I will learn more about the considerations around this. Once the conversion back is done, I will again try the upgrade and see if it goes a little more smoothly.

**Comment by cecilia-donnelly** [comment]
@jasonaowen, @brainwane and I talked about this in IRC just now.  I'm going to summarize that here in the hope that it helps others understand what's going on.

The two options are to (1) keep jBPM, which means upgrading it or (2) replace jBPM with Java, which involves writing some new code.

@jasonaowen is going to estimate the time it would take to follow each of these paths.  Aside from the time, we will also consider (a) how different workflows are likely to be from state to state and (b) how frequently a state will want to change its workflow(s).  If either or both of these are frequent occurrences then taking the time to upgrade jBPM is probably more worthwhile.

**Comment by cecilia-donnelly** [comment]
For historical purposes, I wanted to note here that @jasonaowen and I also talked about using an Eclipse GUI editor to manage rules.   This would show an interface that allows users to edit text files holding the rules.  Technically speaking, each state would then have its own branch in the repository (probably), unless we did some kind of gitignore->include build step.

**Comment by jasonaowen** [comment]
I was able to set up Eclipse with the Drools 5.5 GUI plugin. I'll document that here, for lack of a better place.

1. Download [jbpm-5.4.0.Final-installer-full.zip](https://sourceforge.net/projects/jbpm/files/jBPM%205/jbpm-5.4.0.Final/jbpm-5.4.0.Final-installer-full.zip/download) and extract it.
1. In the extracted jBPM installer, extract `lib/org.drools.updatesite-5.5.0.Final-assembly.zip`.
1. Install and start Eclipse. I was able to use [the version shipped in Ubuntu 16.04](https://packages.ubuntu.com/xenial/eclipse), which meant a simple `sudo apt install eclipse`.
1. The Drools GUI depends on the GEF libraries. Add the GEF repository to Eclipse:
    1. Open the install dialog box: `Help` > `Install New Software...`.
    1. Open the add repository dialog box: `Add...`.
    1. Name the new repository `GEF`, and set the location to `http://download.eclipse.org/tools/gef/updates/releases/`.
1. Add the Drools repository from the locally extracted Drools zip file:
    1. Open the install dialog box: `Help` > `Install New Software...`.
    1. Open the add repository dialog box: `Add...`.
    1. Name the new repository `Drools 5.5`, click `Local`, browse to the directory you extracted  `lib/org.drools.updatesite-5.5.0.Final-assembly.zip` into, and click `OK`.
1. Make sure the newly-added Drools repository is selected.
1. Select the `Drools and jBPM` package (or click `Select All`), and click `Finish`.

Once the GUI is set up, you can make a new Drools project and import the files in `cms-business-process/src/main/resources/`. There are a bunch of errors around missing imports, since it doesn't have access to the associated Java code. I haven't tried yet to make it fully functional, particularly the rules, since the rules editor seems to merely be a text editor with syntax highlighting. The BPMN editor is more interesting, and I might play more with that at some point.

**Comment by brainwane** [comment]
Once this is implemented, please update `psm-app/docs/userhelp/source/service-admin-help.rst` to explain how to use the functionality.

**Comment by jvasile** [comment]
Cecilia, Jason, Dan and James had a call about this.  We want to use JBoss's Drools editing features, not build our own.  We also don't want that stuff tightly coupled to anything else we're deploying.  We decided to explore running the editor as a separate service targeting just one aspect of one provider (Audiologist).  That will tell us more about how to get this moving.

A few considerations: 

 * The editor requires upgrading JBPM, which is a pain.  Running it as a separate microservice decouples the versioning of editor dependencies from the rest of PSM.
 * Drools rules are all over the place in the codebase, so we need to put some work into centralizing that stuff.
 * Some drools changes require corresponding database changes.  That will need to be automated, which might be a challenge.
 * If we run editing as a microservice, we can provide that service to other parts of MMIS that use Drools, which could help simplify things for other aspects of MMIS.

**Comment by jasonaowen** [comment]
To clarify, running jBPM/Drools as a separate service is necessary to present its UI. I had said we needed to upgrade jBPM because the standalone jBPM application that matches version we've embedded is not compatible with WildFly 10.

The good idea that @jvasile suggested today is to run a separate, older application server that is compatible with the standalone application that matches our version of jBPM/Drools. This could be WildFly or even JBoss, depending on what works.

**Comment by jasonaowen** [comment]
> - Drools rules are all over the place in the codebase, so we need to put some work into centralizing that stuff.
> - Some drools changes require corresponding database changes. That will need to be automated, which might be a challenge.

The rules themselves are well-consolidated in [`psm-app/cms-business-process/src/main/resources`](https://github.com/OpenTechStrategies/psm/tree/master/psm-app/cms-business-process/src/main/resources). However, supporting data for some of those rules lives in the database and the UI. For example:

There is [a Drools rule that requires Audiologists to have an Audiologist License](https://github.com/OpenTechStrategies/psm/blob/59248b624c9935b19e8ff29184a32e899e77effd/psm-app/cms-business-process/src/main/resources/cms.screening.drl#L95-L115). However, the UI will only allow a provider to upload an Audiologist License if there is also [seed data that links](https://github.com/OpenTechStrategies/psm/blob/59248b624c9935b19e8ff29184a32e899e77effd/psm-app/db/seed.sql#L860) an [Audiologist provider type](https://github.com/OpenTechStrategies/psm/blob/59248b624c9935b19e8ff29184a32e899e77effd/psm-app/db/seed.sql#L151) with an [Audiologist License type](https://github.com/OpenTechStrategies/psm/blob/59248b624c9935b19e8ff29184a32e899e77effd/psm-app/db/seed.sql#L547).

**Comment by jvasile** [comment]
Thanks for the clarifications, @jasonaowen.

**Comment by slifty** [comment]
Weekend update (cc @cecilia-donnelly) -- I have set up an instance of jboss 7.1.1 on a separate machine and it is running Guvnor with Drools 5.5.0.  I will share a URL privately, as it is not a particularly secured instance.

Some decisions we should try to make in the next day or so:

1. How do we integrate this UI to our system (e.g. what are the integration points)
2. Do we truly want to use these versions of things.

Item 1 will require some additional research still, and a conversation between myself and @jasonaowen to think through the stones that need to be turned.

Item 2 the major thing to think about is that using JBPM 5.4 and Drools 5.5 means we are forced to use an old versions of several things.  I have not tried turning the knobs to see how recent I can go without breaking things, but the version of jboss that is used by default in the official installer (7.1.1) is unsupported and unpatched (so it could very well have known zero days).

**Comment by jasonaowen** [comment]
Thank you for setting that up, @slifty ! I'll take a look at importing our rules into that instance and having the PSM use it as a knowledge base. I'd be happy to talk through that with you today or tomorrow.

For item 2, I completely agree that there are a lot of security concerns. I think some of them can be mitigated through configuration, such as not allowing access from any host other than the PSM's. My goal in this is to have it be a stepping stone to upgrading to a more recent version, so hopefully this will be a short-lived setup; maybe (fingers crossed) it won't even ever have to leave a feature branch. That said, if it's not too much of a hassle, we should probably talk about figuring out the last supported application server for that version, and maybe setting up some firewall rules to limit its vulnerability.

**Comment by jasonaowen** [comment]
Yesterday I was able to start using the instance @slifty set up to act as a rules repository for the embedded Drools in the PSM. There is already supporting code to do that, so the changes were few:
- change the value of `rules.embedded` from `Y` to `N` in `cms.properties`
- add location and authentication info to `ExternalSourcesRules.xml`, `ScreeningRules.xml`, and `ValidationRules.xml`. One such diff looks like:

```diff
-         <resource source='http://localhost:8080/cms-rules/org.drools.guvnor.Guvnor/package/cms.externalsources/LATEST.drl' type='DRL' />
+         <resource source='http://jboss7.example.com:8080/drools-guvnor/org.drools.guvnor.Guvnor/package/cms.externalsources/LATEST.drl'
+                   type='DRL'
+                   basicAuthentication="enabled"
+                   username="username"
+                   password="password"
+         />
```

I was pleased by how well that worked!

**Comment by jasonaowen** [comment]
Today I continued trying to move our jBPM workflow (stored in [EnrollmentProcess.bpmn](https://github.com/OpenTechStrategies/psm/blob/52ebd2984c00c4f7a33e41414e24fd3e4b0965af/psm-app/cms-business-process/src/main/resources/EnrollmentProcess.bpmn)) into that instance, to no success. We appear to be running into [Bug 768375 - Guvnor jBPM Designer does not import bpmn2 files](https://bugzilla.redhat.com/show_bug.cgi?id=768375), which is still open six years later. :/

Since importing our workflow appears to be blocked, I started reproducing it by hand. I had [previously set up Eclipse](https://github.com/OpenTechStrategies/psm/issues/274#issuecomment-318401081), so I used that view of our current workflow to begin recreating it in the web UI. This is turning out to be somewhat useful, as there is a lot more detail to a workflow than is shown in a rendered image, including things such as executable Java code that tells jBPM which links to follow in the graph. It's also kind of slow going, as I am both learning how our workflow works and figuring out the differences between the Eclipse and web UIs.

I'll be out for the next several days. If someone else with access wants to poke at importing (automatically or by hand) our workflow into Guvnor, they are more than welcome to. :)

**Comment by slifty** [comment]
On my end I pushed the script to create and deploy the service on an REHL instance over at #371 

Jason lets find a window to chat in the morning tomorrow to talk through how I might be able to help with that bug.  That ticket makes it sound possible that it's a function of the deployment environment rather than the code base itself.

**Comment by cecilia-donnelly** [comment]
Closing this in favor of #385 and #386.  We can (of course) reopen as needed.
